### ISSUE TITLE
[FEA] could the parquet scan code avoid acquiring the semaphore for an empty batch?

### ISSUE THREAD

**Issue by abellina** [NVIDIA/spark-rapids#4392]
The code here: https://github.com/NVIDIA/spark-rapids/blob/branch-22.02/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala#L716 acquires the GPU semaphore if the scanned batch is empty. This is the cause of an odd pattern we are seeing in the NVTX ranges (and confirms a hunch @jlowe had when he noticed the extra acquire), that can show expensive waits over what looks like to be a noop, depending on the plan:

![Screenshot from 2021-12-20 08-29-30](https://user-images.githubusercontent.com/1901059/146782917-f9fcf6a1-45c2-44da-b9c9-d036fbc31991.png)

In this query (NDS q64), the scanned batches are going into a filter for them to get shuffled:
![Screenshot from 2021-12-20 08-33-41](https://user-images.githubusercontent.com/1901059/146783665-efa9d8a1-afde-4154-8170-388112cf2909.png)

In this case ideally we should acquire the GPU if the filter wants to go to the GPU. This is not easy to do at first glance, and it may need some discussion on how best to tackle it, likely at planning time (i.e. will my downstream exec always produce a GPU batch, then always acquire on empty).

**Comment by jlowe** [comment]
> In this case ideally we should acquire the GPU if the filter wants to go to the GPU.

I suspect the Parquet scan is not returning an empty batch, rather `hasNext` should be returning false indicating there are no batches.  The filter should not try to go to the GPU if there are no batches.

The original reason to have the Parquet scan always grab the semaphore, even if there are no batches, is to cover the case where an exec node can produce data _even when there's no input_.  Counting aggregates are a classic case of this, where it needs to produce an output even when there are no inputs (i.e.: emitting a zero value for the count).

So removing this requires finding all the places in the code that, like a counting aggregate, will try to perform GPU operations despite receiving no input batches.  Originally we didn't want to try to track all of them down (and remember to add acquisition on any new operations added that fall in the same category).  If we're seeing significant performance loss due to excessive over-provisioning on the input files (i.e.: lots of first stage tasks receiving no input) then we may need to reconsider that decision.

**Comment by revans2** [comment]
So what is the opportunity cost of acquiring the semaphore? Could that task actually do any real useful work instead of waiting on the semaphre? The only thing it could do in this case is to buffer data for the next batch. I can see it possibly improving things, but I could also see it slowing things down.  If you are on cloud storage and a single connection is slow, but you can get good bandwidth in aggregate, then yes this is likely to help. If the I/O is actually limited by network or physical disks starting more I/O while others are happening might even slow down the query in aggregate. I agree that this is something that would be nice to do. And I think the only place we have that outputs something even when there is no input is a reduction.  So it should be doable. I just don't think it is going to have any noticeable impact to performance in most cases.

**Comment by jlowe** [comment]
> The only thing it could do in this case is to buffer data for the next batch.

I could see this helping in cases where the degenerate tasks that now wait for the semaphore needlessly are the trailing tasks in a stage and waiting for the semaphore because they are competing with tasks in another stage running in parallel.  By allowing these degenerate tasks to complete early, we allow the subsequent stage to be scheduled and thus start working on processing their shuffle inputs, coalescing the batches on the host before they ultimately need to wait for the GPU semaphore.  This is essentially another spin on the "buffer data for the next batch" scenario described above, but it doesn't involve cloud storage.

**Comment by abellina** [comment]
For NDS at 3TB this looks to be a wash, with about 1/2 the queries being faster and the other 1/2 slower (and by the same amounts), overall it was about 9 seconds faster total. I tried removing the acquire from the specific spot here: https://github.com/NVIDIA/spark-rapids/blob/branch-22.02/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala#L716. 

That said, when looking at a specific stage you can absolutely see tasks go from multi-seconds in duration, to a few millis. I think another scenario where we'd see benefits is if we had a lot more tasks than slots with skew. I.e. if half the tasks return a lot of data and are expensive, and the other half are very cheap, then we may not get to the next expensive task because we are holding onto the GPU from the previous wave. That said, I see no real benefit in NDS right now.

**Comment by revans2** [comment]
Like I said I think this is a good change. And I think it lays the groundwork for possibly other changes that we could make. So I am +1 for putting it in assuming we don't see a huge regression on any one query.

**Comment by jlowe** [comment]
Note that we should catch _all_ the places where this could occur, i.e.: the various multi-file readers, coalesce readers, single-file readers, etc. across all input types (Parquet, ORC, CSV, etc.)  Also do we need to worry about this for shuffle, i.e.: is there a case where shuffle read could grab the semaphore for no/empty batches for a task?

**Comment by abellina** [comment]
The original issue in this ticket was handled by https://github.com/NVIDIA/spark-rapids/pull/4476, I am closing this and opening another issue to track @jlowe's last comment.
